### PR TITLE
implement caveat support for spanner

### LINKED ISSUE #917: support caveats in Spanner datastore
Tracking: https://github.com/authzed/spicedb/issues/386
Part of https://github.com/authzed/spicedb/issues/835

Add support in `spanner` datastore to support caveats

### PR DESCRIPTION
**Pull request by ecordell**
Fixes #917

### CODE REVIEW

File: internal/datastore/spanner/watch.go [hunk 1]
```diff
@@ -102,10 +106,22 @@ func (sd spannerDatastore) loadChanges(
 			&tpl.Subject.Namespace,
 			&tpl.Subject.ObjectId,
 			&tpl.Subject.Relation,
+			&caveatName,
+			&caveatCtx,
 		)
 		if err != nil {
 			return err
 		}
+		if caveatName.Valid && caveatCtx.Valid {
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:115]: hrm, why checking `caveatCtx.Valid` though? It's entirely possible to have a caveated relationship with a null context. May be worth checking if our tests exercise getting updates for both relationships with caveat but null context and caveat with non-null context

File: internal/datastore/spanner/watch.go [hunk 2]
```diff
@@ -102,10 +106,22 @@ func (sd spannerDatastore) loadChanges(
...
+			caveatStruct, err := structpb.NewStruct(caveatCtx.Value.(map[string]interface{}))
```
Comment (vroldanbet) [inline internal/datastore/spanner/watch.go:116]: Is the `JSON` type from Spanner a map[string]any under the hood, so we know this type assertion won't panic?

also super nit 😄 
```suggestion
			caveatStruct, err := structpb.NewStruct(caveatCtx.Value.(map[string]any))
```

File: internal/datastore/spanner/caveat.go [hunk 3]
```diff
@@ -3,23 +3,122 @@ package spanner
...
+
+	keyset := spanner.AllKeys()
+	if len(caveatNames) > 0 {
+		keys := make([]spanner.Key, 0, len(caveatNames))
+		for _, n := range caveatNames {
+			keys = append(keys, spanner.Key{n})
+		}
+		keyset = spanner.KeySetFromKeys(keys...)
+	}
+	iter := sr.txSource().Read(
```
Comment (vroldanbet) [inline internal/datastore/spanner/caveat.go:54]: Is Spanner ordering by any key implicitly? Would it be possible for the output to have a non-deterministic order?

File: internal/datastore/spanner/caveat.go [hunk 4]
```diff
@@ -3,23 +3,122 @@ package spanner
...
+
+	return caveats, nil
 }
 
 func (rwt spannerReadWriteTXN) WriteCaveats(caveats []*core.CaveatDefinition) error {
-	return fmt.Errorf("unimplemented caveat support in datastore")
+	_, span := tracer.Start(rwt.ctx, "WriteCaveats")
+	defer span.End()
+
+	names := map[string]struct{}{}
```
Comment (vroldanbet) [inline internal/datastore/spanner/caveat.go:86]: why not relying on the database to impose key uniqueness instead? I can understand it being faster given it does not have to hit the DB, but we still need to rely on the database anyway given an overlapping transaction could attempt to store the same caveat?

File: internal/datastore/spanner/reader.go [hunk 5]
```diff
@@ -86,14 +87,28 @@ func queryExecutor(txSource txFactory) common.ExecuteQueryFunc {
...
 				&nextTuple.ResourceAndRelation.Namespace,
 				&nextTuple.ResourceAndRelation.ObjectId,
 				&nextTuple.ResourceAndRelation.Relation,
 				&nextTuple.Subject.Namespace,
 				&nextTuple.Subject.ObjectId,
 				&nextTuple.Subject.Relation,
+				&caveatName,
+				&caveatCtx,
 			)
+			if caveatName.Valid && caveatCtx.Valid {
```
Comment (vroldanbet) [inline internal/datastore/spanner/reader.go:102]: Same here, why enforcing caveat context needs to be non-null? I'm surprised no test fails, we may have a gap.

File: internal/datastore/spanner/readwrite.go [hunk 6]
```diff
@@ -177,7 +186,16 @@ func deleteWithFilter(ctx context.Context, rwt *spanner.ReadWriteTransaction, fi
 
 func upsertVals(r *core.RelationTuple) []interface{} {
 	key := keyFromRelationship(r)
-	return append(key, spanner.CommitTimestamp)
+	key = append(key, spanner.CommitTimestamp)
+	if r.Caveat != nil && len(r.Caveat.CaveatName) > 0 {
+		key = append(key, r.Caveat.CaveatName)
+		if r.Caveat.Context != nil {
```
Comment (vroldanbet) [inline internal/datastore/spanner/readwrite.go:192]: shouldn't we appending a `nil` context if it's empty? we seem to be doing it in the else block below?

File: internal/datastore/spanner/caveat.go (hunk 3 above)
Comment (josephschorr) [reply internal/datastore/spanner/caveat.go:54]: Is order implied in this interface though?

File: internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"context"
+
+	"google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+const (
+	createCaveatTable = `CREATE TABLE caveat (
+    	name STRING(1024),
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go:11]: We should add the limits into the core definitions as well; I'll prep a PR

[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go:11]: Just checked: its limit to 128 bytes in the core protos

File: internal/datastore/spanner/reader.go (hunk 5 above)
Comment (josephschorr) [reply internal/datastore/spanner/reader.go:102]: Isn't the context nil when it is unspecified?

File: internal/datastore/spanner/readwrite.go [hunk 8]
```diff
@@ -128,7 +126,9 @@ func deleteWithFilter(ctx context.Context, rwt *spanner.ReadWriteTransaction, fi
 	rel := core.RelationTuple{
 		ResourceAndRelation: &core.ObjectAndRelation{},
 		Subject:             &core.ObjectAndRelation{},
+		Caveat:              &core.ContextualizedCaveat{},
```
Comment (josephschorr) [inline internal/datastore/spanner/readwrite.go:129]: Ideally the caveat starts as `nil`

File: internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go:11]: a limit [already exists](https://github.com/authzed/spicedb/blob/376dbfd3385c73f33545c479daff2c4c9e4b0d03/proto/internal/core/v1/core.proto#L44) for the definition in `core.proto`. That's where the 4KB limit comes from. I was bringing up if we should also add those limits in the SQL migrations. I'd argue that if we have limits in the API, and we have the need to increase it, we can avoid a migration.

File: internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    	definition BYTES(MAX) NOT NULL,
+    	timestamp TIMESTAMP NOT NULL OPTIONS (allow_commit_timestamp=true)
+	) PRIMARY KEY (name)`
+
+	addRelationshipCaveatName = `ALTER TABLE relation_tuple
```
[the remark this answers is not in the document]
Comment (ecordell) [reply internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go:16]: I couldn't find any syntax that would let me add multiple columns in one statement, but these all get sent in 1 DDL request to spanner.

File: internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ecordell) [reply internal/datastore/spanner/migrations/zz_migration.0003_add_caveats.go:11]: > I'd argue that if we have limits in the API, and we have the need to increase it, we can avoid a migration.

This seems reasonable to me. From [what I can tell](https://stackoverflow.com/questions/45964937/performance-difference-for-stringmax) there's no performance penalty to make these `STRING(MAX)` so doing that is probably the most flexible.

File: internal/datastore/spanner/caveat.go (hunk 4 above)
Comment (ecordell) [reply internal/datastore/spanner/caveat.go]: Not sure why I did this - removed it.

Comment (vroldanbet) [reply internal/datastore/spanner/caveat.go:87]: Idempotency FTW 😄

File: internal/datastore/spanner/caveat.go (hunk 3 above)
Comment (ecordell) [reply internal/datastore/spanner/caveat.go:55]: I didn't think order mattered here?

Comment (vroldanbet) [reply internal/datastore/spanner/caveat.go:55]: Well, I guess it makes API consumption easier - I think I added it in the first place for some tests, I'd need to revisit if that requirement still holds. CRDB/PSQL implementation do certainly have the order by clause.

**Review by vroldanbet — APPROVED** [review]

**Pull request merged**